File: scheduler/token_bucket.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# @Author: darrenxyli <www.darrenxyli.com>
# @Date:   2014-10-30 00:36:51
# @Last Modified by:   darrenxyli
# @Last Modified time: 2014-10-30 00:37:00

import time
try:
    import threading as _threading
except ImportError:
    import dummy_threading as _threading

class Bucket(object):
    '''
    traffic flow control with token bucket
    '''

    update_interval = 30
    def __init__(self, rate=1, burst=None):
        self.rate = float(rate)
        if burst is None:
            self.burst = float(rate)*10
        else:
            self.burst = float(burst)
        self.mutex = _threading.Lock()
        self.bucket = self.burst
        self.last_update = time.time()

    def get(self):
        now = time.time()
        if self.bucket >= self.burst:
            self.last_update = now
            return self.bucket
        bucket = self.rate * (now - self.last_update)
        self.mutex.acquire()
        if bucket > 1:
            self.bucket += bucket
            if self.bucket > self.burst:
                self.bucket = self.burst
            self.last_update = now
        self.mutex.release()
        return self.bucket

    def set(self, value):
        self.bucket = value

    def desc(self, value=1):
        self.bucket -= value

File: scheduler/token_bucket.go
package scheduler

import (
	"sync"
	"time"
)

// Bucket flow control
type Bucket struct {
	updateInterval uint
	rate           float32
	burst          float32
	avaliableToken float32
	lock           *sync.Mutex
	lastUpdate     float32
}

// NewBucket is defalut constructor
func NewBucket(rateP float32, burstP float32) *Bucket {
	now := float32(time.Now().Second())
	return &Bucket{
		// Refresh interval
		updateInterval: 30,
		// Regenerate token rate
		rate: rateP,
		// Limit size of bucket
		burst: burstP,
		// Lock
		lock: new(sync.Mutex),
		// Avaliable tokens in bucket now
		avaliableToken: burstP,
		// The last update time
		lastUpdate: now,
	}
}

// Get the avaliable token in Bucket
func (buc *Bucket) Get() float32 {
	now := float32(time.Now().Second())
	if buc.avaliableToken >= buc.burst {
		buc.lastUpdate = now
		return buc.avaliableToken
	}

	// Generate now token
	bucket := buc.rate * (now - buc.lastUpdate)
	buc.lock.Lock()
	if bucket > 1 {
		buc.avaliableToken += bucket
		if buc.avaliableToken > buc.burst {
			buc.avaliableToken = buc.burst
		}
		buc.lastUpdate = now
	}
	buc.lock.Unlock()

	// Give avaliable token
	return buc.avaliableToken
}

// Set numbers of token in bucket
func (buc *Bucket) Set(size float32) {
	buc.avaliableToken = size
}

// Desc to decrease the avaliable token
func (buc *Bucket) Desc(n float32) {
	buc.avaliableToken -= n
}
